src/source/image_source.js

```diff
@@ -140,40 +141,43 @@ class ImageSource extends Evented {
     }
 
     _setTile(tile) {
-        this.tile = tile;
+        this.tiles[tile.coord.w] = tile;
         const maxInt16 = 32767;
         const array = new RasterBoundsArray();
         array.emplaceBack(this._tileCoords[0].x, this._tileCoords[0].y, 0, 0);
         array.emplaceBack(this._tileCoords[1].x, this._tileCoords[1].y, maxInt16, 0);
         array.emplaceBack(this._tileCoords[3].x, this._tileCoords[3].y, 0, maxInt16);
         array.emplaceBack(this._tileCoords[2].x, this._tileCoords[2].y, maxInt16, maxInt16);
 
-        this.tile.buckets = {};
+        tile.buckets = {};
 
-        this.tile.boundsBuffer = Buffer.fromStructArray(array, Buffer.BufferType.VERTEX);
-        this.tile.boundsVAO = new VertexArrayObject();
+        tile.boundsBuffer = Buffer.fromStructArray(array, Buffer.BufferType.VERTEX);
+        tile.boundsVAO = new VertexArrayObject();
     }
 
     prepare() {
-        if (!this.tile || !this.image) return;
+        if (this.tiles.length === 0 || !this.image) return;
         this._prepareImage(this.map.painter.gl, this.image);
     }
 
     _prepareImage(gl, image, resize) {
-        if (this.tile.state !== 'loaded') {
-            this.tile.state = 'loaded';
-            this.tile.texture = gl.createTexture();
-            gl.bindTexture(gl.TEXTURE_2D, this.tile.texture);
-            gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
-            gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
-            gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
-            gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
-            gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, gl.RGBA, gl.UNSIGNED_BYTE, image);
-        } else if (resize) {
-            gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, gl.RGBA, gl.UNSIGNED_BYTE, image);
-        } else if (image instanceof window.HTMLVideoElement || image instanceof window.ImageData || image instanceof window.HTMLCanvasElement) {
-            gl.bindTexture(gl.TEXTURE_2D, this.tile.texture);
-            gl.texSubImage2D(gl.TEXTURE_2D, 0, 0, 0, gl.RGBA, gl.UNSIGNED_BYTE, image);
+        for (const w in this.tiles) {
+            const tile = this.tiles[w];
```

Comment:
OK, I took a stab at sharing the texture between tiles -- it passes tests, but I could definitely use a review to see if the approach makes sense. All the image/texture binding stuff just looks like a magic incantation to me.